### PR TITLE
[fsdp,megatron,vllm,trainer,algo] feat: On-Policy Distillation 

### PR DESCRIPTION
**Pull request by JacobHelwig**
# What does this PR do?

Adds on-policy distillation support across FSDP and Megatron backends.

Collaboration with @wuxibin89, including design guidance, restructuring, and added support for parallelism.

Supports: 

- FSDP and Megatron engines 
- top-k distillation loss and KL estimator distillation losses 
- Supervised and policy-gradient-style updates
- Teacher logprobs computation using a vLLM teacher server
- LLM and VLM distillation
- FSDP sequence parallel
- Megatron context parallel and tensor parallel

## Losses 

1. top-k distillation loss: **forward** KL estimated using top-k logits **from teacher**. 
2. KL estimator distillation losses: **reverse** KL estimated using only the log prob for the sampled token via the same estimators used by the reference model (e.g., k1, k3)  

## Updates

1. Supervised: distillation loss is directly backpropagated, as in https://arxiv.org/abs/2306.13649
2. Policy gradient: negative distillation loss is used as a reward, as in https://thinkingmachines.ai/blog/on-policy-distillation/

# Test

- LLM distillation with FSDP: `examples/on_policy_distillation_trainer/run_qwen_gsmk8k.sh`
- VLM distillation with FSDP: `examples/on_policy_distillation_trainer/run_qwen3_vl_geo3k.sh`
- LLM distillation with megatron: `examples/on_policy_distillation_trainer/run_qwen_gsmk8k_megatron.sh`. 

### LLM Distillation 

These experiments compare 3 training runs with student model Qwen2.5-0.5B using `examples/on_policy_distillation_trainer/run_qwen_gsmk8k.sh`:

1. Forward top-k KL with Qwen2.5-3B-Instruct teacher (gold)
2. Forward top-k KL with Qwen2.5-7B-Instruct teacher (green)
3. k3 estimator KL with Qwen2.5-7B-Instruct teacher (red)

#### GSM8K eval acc

image

#### GSM8K train acc

image

#### Distillation loss

image

### VLM Distillation 

- Data: Geometry3K
- Student: Qwen3-VL-2B-Instruct
- Teacher: Qwen3-VL-4B-Instruct
- OPD algo: k1 KL estimator as reward with policy gradient loss

#### Geo3K eval acc

image

#### Geo3K train acc

image

#### Distillation loss

image

## LLM Distillation: Top-k training stability

Clamping the top-k forward KL loss was needed for training stability. These experiments compare 3 types of clamping: 

1. No clamping (grey)
2. Clamping the distillation loss to a maximum value of 10 (blue)
3. Clamping the student and teacher log probs to a minimum value of -10 (gold)

### Distillation loss

image

### GSM8K eval acc

image

### GSM8K train acc

image

## LLM Distillation: Policy-gradient results

While the VLM results in this PR use the k1 KL estimator with policy gradient updates, all LLM distillation results outside of this section rely on supervised updates. LLM distillation with policy gradient updates are validated in this section:

1. Forward top-k KL with supervised update (green)
2. k1 estimator KL with policy gradient update (purple)
3. k3 estimator KL with supervised update (red)

While purple seems best, it also is generating responses that exceed the maximum response length of 512.

### Distillation loss

image

### GSM8K eval acc

image

### GSM8K train acc

image

### Response length

image

## LLM Distillation: Megatron 

To verify parity of megatron engine with FSDP, these experiments compare 3 training runs with student model Qwen2.5-0.5B:

1. Forward top-k KL with Qwen2.5-7B-Instruct teacher + clamping log probs to minimum value of -10.0 (teal)
2. Forward top-k KL with Qwen2.5-3B-Instruct teacher + clamping log probs to minimum value of -10.0 (red)
3. Forward top-k KL with Qwen2.5-3B-Instruct teacher + clamping loss to maximum value of 10.0 (blue)
4. k3 estimator reverse KL with Qwen2.5-7B-Instruct teacher + clamping loss to maximum value of 10.0 (green)

The solid line uses megatron engine with TP=2, the dotted line uses FSDP.

### GSM8K Eval Acc.

image

### GSM8K Train Acc.

image

### Distillation Loss

image

### Grad Norm

image 

## LLM Distillation: Note on reverse KL

Initially, this PR included top-k reverse KL and top-k Jensen-Shannon divergences (JSD interpolates between forward and reverse KL). For the student distribution $q$ and teacher distribution $p$, the top-k reverse KL is given by

$$
KL_{\text{top-}k}(q||p) = \sum_i \bf{1}(q_i\in \text{top-}k)q_i\log\frac{q_i}{p_i}.
$$ 

Unfortunately, this was unstable. The reason is because one way to make this loss small is to make $q_i$ as small as possible for all $q_i \in \text{top}-k$. This can be seen from the logs tracking the amount of mass captured in the top-$k$ probabilities:

image

## LLM Distillation: Ablation: performance with more lenient parser

Note that the only loss used is the distillation loss (no rewards for correctness on GSM8K). Any increase in the logged rewards=GSM8k accuracy are an indirect result of minimizing the distillation loss. The reason that the base model has Pass@1~=0 is because the default GSM8k answer formatting (`#### 42`) is OOD for the model. The base model is answering the questions correctly, but using incorrect formatting, so none of the answers can be parsed. The base model can be evaluated using a reward function that is more lenient on formatting by adding the following to the script:

```bash
...
    reward_model.reward_manager=remote \
    custom_reward_function.path=tests/experimental/reward_loop/reward_fn.py \
    custom_reward_function.name=compute_score_math_verify \
    trainer.val_only=True
``` 

The results are:

```bash
(TaskRunner pid=904198) ("Initial validation metrics: {'val-aux/openai/gsm8k/reward/mean@1': "
(TaskRunner pid=904198)  "np.float64(0.31766489764973466), 'val-core/openai/gsm8k/acc/mean@1': "
(TaskRunner pid=904198)  "np.float64(0.31766489764973466), 'val-aux/num_turns/min': np.int32(2), "
(TaskRunner pid=904198)  "'val-aux/num_turns/max': np.int32(2), 'val-aux/num_turns/mean': "
(TaskRunner pid=904198)  'np.float64(2.0)}')
```

# Design & Code Changes

- Teacher workers are used in the agent loop, similar to the generative reward model: after a student worker finishes its rollout, the teacher worker obtains logprobs
- In the initial version of this PR (#4897), requests were submitted to the `vLLMHttpServer` via the `v1/completions` endpoint, which does not support multi-modal data. While `v1/chat/completions` does support multi-modal inputs, text must be passed as raw text instead of token IDs, preventing exact scoring of student generations since `student gen IDs -> student gen text -> teacher input IDs via v1/chat/completions tokenization` will not always give `student gen IDs == teacher input IDs` (https://vllm.ai/blog/agent-lightning). 
- This PR instead follows a path similar to how rollout replicas directly call the `generate` method on the `vLLMHttpServer`. This enables multi-modal inputs while representing text as token IDs. Requests to the teacher server now call the newly-added `compute_logprobs` method of `vLLMHttpServer`.

### CODE REVIEW

File: verl/workers/engine/megatron/transformer_impl.py [hunk 1]
```diff
@@ -818,7 +835,8 @@ def logits_processor(logits, label, temperature):
                     ret["entropy"] = entropy
                 else:
                     logits_bak = logits
-
+                if self.need_logits_for_distillation:
+                    ret["logits"] = logits.clone()
```
Comment (wuxibin89) [inline verl/workers/engine/megatron/transformer_impl.py]: This `.clone` seems redundant?

File: verl/trainer/distillation/megatron/losses.py [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+    torch.distributed.all_reduce(
+        sum_exp_logits,
+        op=torch.distributed.ReduceOp.SUM,
+        group=get_tensor_model_parallel_group(),
+    )
+    log_sum_exp_logits = sum_exp_logits.log()
+    return vp_logits - log_sum_exp_logits.unsqueeze(dim=-1)
+
+
+class _VocabParallelKLDivergence(torch.autograd.Function):
```
Comment (wuxibin89) [inline verl/trainer/distillation/megatron/losses.py:58]: Not consider context parallel, I can support it.

File: verl/workers/engine/megatron/transformer_impl.py (hunk 1 above)
Comment (JacobHelwig) [reply verl/workers/engine/megatron/transformer_impl.py]: It's because the VP cross entropy function (https://github.com/JacobHelwig/verl/blob/52fa0cd3e8bf16903a508f14de075cff8932687e/verl/workers/engine/megatron/transformer_impl.py#L840) modifies the logits in-place and makes them all extremely small

**Comment by vermouth1992** [comment]
LGTM

**Comment by vermouth1992** [comment]
Please fix conflict
